crypto/fipsmodule/rand/snapsafe_detect.c

```diff
@@ -0,0 +1,153 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR ISC
+
+#include <openssl/crypto.h>
+
+#include "snapsafe_detect.h"
+
+#if defined(OPENSSL_LINUX)
+#include "../delocate.h"
+#include <fcntl.h>
+#include <stdlib.h>
+#include <sys/mman.h>
+#include <unistd.h>
+
+/* Snapsafety state */
+#define SNAPSAFETY_STATE_FAILED_INITIALISE 0x00
+#define SNAPSAFETY_STATE_SUCCESS_INITIALISE 0x01
+#define SNAPSAFETY_STATE_NOT_SUPPORTED 0x02
+
+DEFINE_STATIC_ONCE(aws_snapsafe_init)
+DEFINE_BSS_GET(const char *, sgc_file_path)
+DEFINE_BSS_GET(volatile uint32_t *, sgc_addr)
+DEFINE_BSS_GET(int, snapsafety_state)
+
+/* aws_snapsafe_check_kernel_support returns 1 if the special sysgenid device
+ * file exists and 0 otherwise. */
+static int aws_snapsafe_check_kernel_support(void) {
+  /* This file-exist method is generally brittle. But for our purpose, this
+   * should be more than fine. */
+  if (access(*sgc_file_path_bss_get(), F_OK) != 0) {
+    return 0;
+  }
+  return 1;
+}
+
+static void do_aws_snapsafe_init(void) {
+
+#if defined(AWSLC_SYSGENID_PATH)
+  *sgc_file_path_bss_get() = AWSLC_SYSGENID_PATH;
+#else
+  *sgc_file_path_bss_get() = "/dev/sysgenid";
+#endif
+  *sgc_addr_bss_get() = NULL;
+
+  if (aws_snapsafe_check_kernel_support() != 1) {
+    *snapsafety_state_bss_get() = SNAPSAFETY_STATE_NOT_SUPPORTED;
+    return;
+  }
+
+  /* While not totally obvious from man pages, mmap actually only allocates
+   * memory on page size boundaries. So just give it that hint. */
+  long page_size = sysconf(_SC_PAGESIZE);
+  if (page_size <= 0) {
+    *snapsafety_state_bss_get() = SNAPSAFETY_STATE_FAILED_INITIALISE;
+    return;
+  }
+
+  if (*sgc_addr_bss_get() == NULL) {
```

Comment:
Isn't this always true since this function sets it to null on line 43? Does it make sense to check if sgc_addr_bss_get is null at the start of the function so this is a lazy init that only happens once?